src/query/api/v1/handler/namespace/schema_test.go

```diff
@@ -0,0 +1,233 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/m3db/m3/src/cluster/kv"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/namespace/kvadmin"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	mainProtoStr = `syntax = "proto3";
+
+package mainpkg;
+
+import "mainpkg/imported.proto";
+
+message TestMessage {
+  double latitude = 1;
+  double longitude = 2;
+  int64 epoch = 3;
+  bytes deliveryID = 4;
+  map<string, string> attributes = 5;
+  ImportedMessage an_imported_message = 6;
+}
+`
+	importedProtoStr = `
+syntax = "proto3";
+
+package mainpkg;
+
+message ImportedMessage {
+  double latitude = 1;
+  double longitude = 2;
+  int64 epoch = 3;
+  bytes deliveryID = 4;
+}
+`
+	testSchemaJQ = `
+{
+  name: "testNamespace",
+  msgName: "mainpkg.TestMessage",
+  protoName: "mainpkg/test.proto",
+  protoMap: 
+    {
+      "mainpkg/test.proto": $file1,
+      "mainpkg/imported.proto": $file2
+    }
+}
+`
+)
+
+func genTestJson(t *testing.T) string {
+	tempDir, err := ioutil.TempDir("", "schema_deploy_test")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	file1 := filepath.Join(tempDir, "test.proto")
+	file2 := filepath.Join(tempDir, "imported.proto")
+	require.NoError(t, ioutil.WriteFile(file1, []byte(mainProtoStr), 0644))
+	require.NoError(t, ioutil.WriteFile(file2, []byte(importedProtoStr), 0644))
+
+	jqfile := filepath.Join(tempDir, "test.jq")
+	require.NoError(t, ioutil.WriteFile(jqfile, []byte(testSchemaJQ), 0644))
+
+	file1var := "\"$(<" + file1 + ")\""
+	file2var := "\"$(<" + file2 + ")\""
+	cmd := exec.Command("/bin/sh", "-c", "jq -n --arg file1 "+file1var+
+		" --arg file2 "+file2var+" -f "+jqfile)
+	t.Logf("cmd args are %v\n", cmd.Args)
+	out, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+	t.Logf("generated json is \n%s\n", string(out))
+	return string(out)
+}
+
+func TestGenTestJson(t *testing.T) {
```

Comment:
`--- PASS: TestSchemaDeploy (0.02s)
    schema_test.go:102: cmd args are [/bin/sh -c jq -n --arg file1 "$(</var/folders/vg/khsllfnx1qg8q5jk6ggj774c0000gn/T/schema_deploy_test114605121/test.proto)" --arg file2 "$(</var/folders/vg/khsllfnx1qg8q5jk6ggj774c0000gn/T/schema_deploy_test114605121/imported.proto)" -f /var/folders/vg/khsllfnx1qg8q5jk6ggj774c0000gn/T/schema_deploy_test114605121/test.jq]
    schema_test.go:105: generated json is
        {
          "name": "testNamespace",
          "msgName": "mainpkg.TestMessage",
          "protoName": "mainpkg/test.proto",
          "protoMap": {
            "mainpkg/test.proto": "syntax = \"proto3\";\n\npackage mainpkg;\n\nimport \"mainpkg/imported.proto\";\n\nmessage TestMessage {\n  double latitude = 1;\n  double longitude = 2;\n  int64 epoch = 3;\n  bytes deliveryID = 4;\n  map<string, string> attributes = 5;\n  ImportedMessage an_imported_message = 6;\n}",
            "mainpkg/imported.proto": "\nsyntax = \"proto3\";\n\npackage mainpkg;\n\nmessage ImportedMessage {\n  double latitude = 1;\n  double longitude = 2;\n  int64 epoch = 3;\n  bytes deliveryID = 4;\n}"
          }
        }`

Reply:
@Haijuncao How do you feel about base64 encoding the files? Then it can still be done from the command line if we want, but it also makes it was easier to manually construct a request if necessary too (`cat schema.proto | base64`) or they can do it in a web browser too etc.

Other than that I like this much more than the previous YAML approach :)

Reply:
I don't see why we want to base64 the file given the jq approach? I prefer jq approach because it is more human friendly.

Reply:
I like the base64 approach because:

1. Its a common way to include files in JSON payloads
2. Its easier for someone to construct a payload manually (I.E if they don't have access / dont want to use `jq`) because its much easier to generate a base64 encoded string that you can use in your terminal then get all the quoting that `jq` is doing correct.


That said I don't have super strong feelings on it, as long as we have a well structured payload like what you have above I'm cool with it